internal/graph/graph.go

```diff
@@ -0,0 +1,138 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package graph
+
+// Graph represents a simple interface for representation
+// of a directed graph.
+// It is assumed that each node in the graph is uniquely
+// identified with an incremental positive integer (i.e. 1, 2, 3...).
+// A value of 0 for a node represents a sentinel error value.
+type Graph interface {
+	// Order returns the total number of nodes in the graph
+	Order() int
+
+	// EdgesFrom returns a list of integers that each
+	// represents a node that has an edge from node u.
+	EdgesFrom(u int) []int
+}
+
+// IsAcyclic uses depth-first search to find cycles
+// in a generic graph represented by Graph interface.
+// If a cycle is found, it returns a list of nodes that
+// are in the cyclic path, identified by their orders.
+func IsAcyclic(g Graph) (bool, []int) {
+	// cycleStart is a node that introduces a cycle in
+	// the graph. Values in the range [0, g.Order()] mean
+	// that there exists a cycle in g.
+	cycleStart := -1
+	info := newCycleInfo(g.Order())
+
+	for i := 1; i < g.Order(); i++ {
+		info.Reset()
+
+		cycleStart = isAcyclic(g, i, info)
+		if cycleStart >= 0 {
+			break
+		}
+	}
+
+	if cycleStart < 0 {
+		return true, nil
+	}
+
+	// compute cycle path using backtrack
+	cycle := []int{cycleStart}
+	curr := cycleStart
+	for {
+		curr = info.nodes[curr].Backtrack
+		cycle = append(cycle, curr)
+		if curr == cycleStart {
+			break
+		}
+	}
+
+	// cycle is reverse-order.
+	i, j := 0, len(cycle)-1
+	for i < j {
+		cycle[i], cycle[j] = cycle[j], cycle[i]
+		i++
+		j--
+	}
+	return false, cycle
+}
+
+// isAcyclic traverses the given graph starting from a specific node
+// using depth-first search using recursion. If a cycle is detected,
+// it returns the node that contains the "last" edge that introduces
+// a cycle.
+// For example, running isAcyclic starting from 1 on the following
+// graph will return 3.
+// 	1 -> 2 -> 3 -> 1
+func isAcyclic(g Graph, u int, info *cycleInfo) int {
+	info.nodes[u].Visited = true
+	info.nodes[u].OnStack = true
+
+	for _, v := range g.EdgesFrom(u) {
+		if !info.nodes[v].Visited {
+			info.nodes[v].Backtrack = u
+			if start := isAcyclic(g, v, info); start >= 0 {
+				return start
+			}
+		} else if info.nodes[v].OnStack {
+			info.nodes[v].Backtrack = u
+			return v
+		}
+	}
+	info.nodes[u].OnStack = false
+	return -1
+}
+
+// cycleNode keeps track of a single node's info for cycle detection.
+type cycleNode struct {
+	Visited   bool
+	OnStack   bool
+	Backtrack int
+}
+
+// cycleInfo contains helpful info for cycle detection.
+type cycleInfo struct {
+	// order is the number of nodes in the graph
+	order int
+
+	// nodes is the information for a given node.
+	nodes []cycleNode
+}
+
+func newCycleInfo(order int) *cycleInfo {
+	return &cycleInfo{
+		order: order,
+		// +1 because 0 is always a sentinel value.
+		nodes: make([]cycleNode, order+1),
```

Comment:
Wait, this is off. The internal graph type and its algorithm doesn't have a concept of a sentinel node.
In Dig's usage of this graph, it happens to put a sentinel node at position 0,
and that node will have no dependencies or dependents,
but that shouldn't leak to this code here?

Reply:
oh, yes. That's correct. I'll correct this.